RecoParticleFlow/PFClusterProducer/src/Cluster3DPCACalculator.cc

```diff
@@ -0,0 +1,132 @@
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/isFinite.h"
+
+#include <cmath>
+#include <unordered_map>
+
+#include "vdt/vdtMath.h"
+
+#include "RecoParticleFlow/PFClusterProducer/interface/PFCPositionCalculatorBase.h"
+#include "DataFormats/ParticleFlowReco/interface/PFRecHitFwd.h"
+#include "DataFormats/ParticleFlowReco/interface/PFRecHit.h"
+
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+
+#include "TPrincipal.h"
+
+class Cluster3DPCACalculator : public PFCPositionCalculatorBase {
+ public:
+  Cluster3DPCACalculator(const edm::ParameterSet& conf) :
+    PFCPositionCalculatorBase(conf),    
+    _pca(new TPrincipal(3,"D")){ 
+  }
+  Cluster3DPCACalculator(const Cluster3DPCACalculator&) = delete;
+  Cluster3DPCACalculator& operator=(const Cluster3DPCACalculator&) = delete;
+
+  void calculateAndSetPosition(reco::PFCluster&);
+  void calculateAndSetPositions(reco::PFClusterCollection&);
+
+ private:
+  std::unique_ptr<TPrincipal> _pca;
+
+  void showerParameters(const reco::PFCluster&, math::XYZPoint&, 
+			math::XYZVector& );
+
+  void calculateAndSetPositionActual(reco::PFCluster&);
+};
+
+DEFINE_EDM_PLUGIN(PFCPositionCalculatorFactory,
+		  Cluster3DPCACalculator,
+		  "Cluster3DPCACalculator");
+
+void Cluster3DPCACalculator::
+calculateAndSetPosition(reco::PFCluster& cluster) {
+  _pca.reset(new TPrincipal(3,"D"));
+  calculateAndSetPositionActual(cluster);
+}
+
+void Cluster3DPCACalculator::
+calculateAndSetPositions(reco::PFClusterCollection& clusters) {
+  for( reco::PFCluster& cluster : clusters ) {
+    _pca.reset(new TPrincipal(3,"D"));
+    calculateAndSetPositionActual(cluster);
+  }
+}
+
+void Cluster3DPCACalculator::
+calculateAndSetPositionActual(reco::PFCluster& cluster) {  
+  if( !cluster.seed() ) {
+    throw cms::Exception("ClusterWithNoSeed")
+      << " Found a cluster with no seed: " << cluster;
+  }  				
+  double cl_energy = 0;  
+  //double cl_time = 0;  
+  //double cl_timeweight=0.0;
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant